### PR TITLE
Update compiler-rt to LLVM 18.1.2

### PR DESCRIPTION
**Pull request by aheejin**
This updates compiler-rt to LLVM 18.1.2.

Things to note:
- https://github.com/llvm/llvm-project/pull/73573 reformatted https://github.com/emscripten-core/emscripten/blob/main/system/lib/compiler-rt/lib/sanitizer_common/sanitizer_linux.cpp, so our diff got somewhat bigger because adding a preprocessor like `ifdef` somewhere means changing the indentation of all inner directives.

- In https://github.com/emscripten-core/emscripten/blob/main/system/lib/compiler-rt/lib/sanitizer_common/sanitizer_linux.cpp,
  ```diff
  - #elif SANITIZER_USE_GETAUXVAL
  -   return getauxval(AT_PAGESZ);
  ```
  These lines were somehow deleted in #11662 for no apparent reason. (It looks this is not used by us) Restored it to reduce the diff between us and the upstream.

- Build libsanitizer_common_rt with `-Wno-format`: https://github.com/emscripten-core/emscripten/pull/21663/commits/67108b2ebafdd6e8f74b89c5adefca9c1e16286d
  Some code that got newly compiled in this version has violations to `__attribute__ ((format))` that crashes compilation without `-Wno-format`. Detailed description of the problem is in https://github.com/emscripten-core/emscripten/pull/21663/commits/67108b2ebafdd6e8f74b89c5adefca9c1e16286d. The simple temporary fix of adding `-Wno-format` is basically what the upstream code currently [does](https://github.com/llvm/llvm-project/blob/26a1d6601d727a96f4301d0d8647b5a42760ae0c/compiler-rt/lib/sanitizer_common/CMakeLists.txt#L223-L224) with `CMakeLists.txt`.

- `ThreadStart`-related changes in https://github.com/emscripten-core/emscripten/blob/main/system/lib/compiler-rt/lib/asan/asan_emscripten.cpp: https://github.com/emscripten-core/emscripten/pull/21663/commits/36ce3d3a574abf70e5a24d5d1281ce6e342532d7
This deals with the following upstream changes:
https://github.com/llvm/llvm-project/commit/4e1b55a781946303202fdf412acb4f151f0474d0
https://github.com/llvm/llvm-project/commit/fd16d465363ecf43c133e881038d87dc06704c0f
by adding `ThreadStartParams` class as https://github.com/llvm/llvm-project/blob/main/compiler-rt/lib/asan/asan_win.cpp did.

Each fix is described in its own commit, except for the fixes required to resolve conflicts when merging our changes, which I wasn't able to commit separately.

### CODE REVIEW

File: tools/system_libs.py [hunk 1]
```diff
@@ -2059,7 +2062,7 @@ class libsanitizer_common_rt(CompilerRTLibrary, MTLibrary):
               'system/lib/compiler-rt/lib',
               'system/lib/libc']
   never_force = True
```
Comment (sbc100) [inline tools/system_libs.py:2064]: Add a comment/TODO here with link the upstream issue?

**Review by sbc100 — APPROVED** [review]
Nice!

File: tools/system_libs.py (hunk 1 above)
Comment (aheejin) [reply tools/system_libs.py:2064]: Not sure if there is a specific upstream issue for this. Will add a todo saying we can remove this after upstream format problems are resolved.

Comment (aheejin) [reply tools/system_libs.py:2064]: Will paste a link to https://github.com/llvm/llvm-project/blob/da675b922cca3dc9a76642d792e882979a3d8c82/compiler-rt/lib/sanitizer_common/CMakeLists.txt#L225-L226 then.

**Review by sbc100 — APPROVED** [review]

**Pull request merged**